crates/ruff_python_formatter/src/expression/string.rs

```diff
@@ -1080,74 +1075,105 @@ fn format_docstring(normalized: &NormalizedString, f: &mut PyFormatter) -> Forma
     write!(f, [source_position(normalized.end()), normalized.quotes])
 }
 
-/// If the last line of the docstring is `content" """` or `content\ """`, we need a chaperone space
-/// that avoids `content""""` and `content\"""`. This does only applies to un-escaped backslashes,
-/// so `content\\ """` doesn't need a space while `content\\\ """` does.
-fn needs_chaperone_space(normalized: &NormalizedString, trim_end: &str) -> bool {
-    trim_end.ends_with(normalized.quotes.style.as_char())
-        || trim_end.chars().rev().take_while(|c| *c == '\\').count() % 2 == 1
-}
-
-/// Format a docstring line that is not the first line
-fn format_docstring_line(
-    line: &str,
-    is_last: bool,
+/// An abstraction for printing each line of a docstring.
+struct DocstringLinePrinter<'ast, 'buf, 'fmt> {
```

Comment:
I think the way it is implemented is fine, considering that `print` takes `lines` as an argument. 

The alternative (that @konstin) suggested is to instead pass `lines` to the `DocstringLinePrinter` constructor and implement `Format` for `DocstringLinePrinter`. I would then probably rename the struct to `DocstringLines`, because it is now a struct that knows how to format itself. 

However, I believe the problem with that is that `Format::format` takes `&self` and not `&mut self`, making it impossible for the printer to update its internal fields (without using internal mutability)

Reply:
Aye. I also expect this abstraction to get mangled and refactored at some point in the future. Namely, I expect that it will become useful to have "look for Python code snippets" logic outside of the context of the Python formatter. That is to say, I don't expect this is its final form. :-)